src/common/snippets/src/pass/insert_loops.cpp

```diff
@@ -5,26 +5,222 @@
+        }
+
+        const auto& inner_loop_begin = op::insertLoopBeginAfterOutputs(body_parameters);
+        const auto& inner_loop_end = op::insertLoopEndBeforeInputs(
+            body_results, inner_loop_begin, inner_work_amount, vector_size,
+            apply_increments, inner_finalization_offsets);
+        // set internal flag to enable scalar vs vector loop optimizations
+        inner_loop_end->has_outer_loop = outer_work_amount > 1;
+        // Due to features of topological sort, some Constants (Scalars) may appear right after Parameters in
+        // sorted ops (so it's between Parameters and LoopBegin). Consequently, ScalarEmitters would be called
+        // outside the Loop, and only the first Loop iteration would yield correct data (assuming the vector reg
+        // assigned to scalar will get corrupted inside the loop body). To avoid such cases, we add control dependency
+        // on LoopBegin to guarantee that the constants are executed inside the Loop.
+        for (const auto& n : body) {
+            if (auto c = std::dynamic_pointer_cast<ov::op::v0::Constant>(n)) {
+                c->add_control_dependency(inner_loop_begin);
+            }
+        }
+
+        if (outer_work_amount > 1) {
+            std::vector<bool> apply_increments = InsertLoops::calculate_outer_apply_increments(body_shapes);
+            std::vector<int64_t> outer_finalization_offsets(body_shapes.size(), 0);
+            const auto& outer_loop_begin = op::insertLoopBegin(body_parameters);
+            op::insertLoopEnd(body_results, outer_loop_begin, outer_work_amount, 1lu,
+                apply_increments, outer_finalization_offsets);
+        }
+    };
+
+    auto op_is_outside_loop = [](const std::shared_ptr<ov::Node>& op) -> bool {
+        if (ov::is_type<ov::op::v0::Parameter>(op) ||
+            ov::is_type<ov::op::v0::Result>(op) ||
+            ov::is_type<op::Buffer>(op))
+            return true;
+        auto& rt = op->get_rt_info();
+        auto outside_rt = rt.find("outside_loop");
+        bool is_outside = false;
+        // If rt info isn't setted it means that op should be inside loop by default
+        if (outside_rt != rt.end()) {
+            is_outside = outside_rt->second.as<bool>();
+        }
+        return is_outside;
+    };
+
+    for (auto iter = ops.begin(); iter < ops.end(); iter++) {
+        const auto op = *iter;
+        // Need to check for that op should be inside or outside loop
+        if (op_is_outside_loop(op)) {
+            continue;
+        }
+
+        // If we meet loopBegin or Brgemm, it means that all previous nodes from ordered body
+        // should be in one body. It's like stop signal
+        const auto& loop_begin = ov::as_type_ptr<op::LoopBegin>(op);
+        const auto& brgemm = ov::as_type_ptr<op::Brgemm>(op);
+        if (loop_begin || brgemm) {
+            if (body.size() > 0) {
```

Comment:
Minor: I've noticed that you don't like `empty()` for some reason 😄 `!body.empty()`?
Here and in lots of other places in this file)